src/openms/source/ANALYSIS/ID/SearchDatabase.cpp

```diff
@@ -0,0 +1,337 @@
+    {return (all_peptides_[l.peptide_index_].peptide_mz_ < all_peptides_[r.peptide_index_].peptide_mz_);});
+  }
+  
+}
+
+void SearchDatabase::updateMembers_()
+{
+  digestor_enzyme_ = param_.getValue("digestor_enzyme").toString();
+  missed_cleavages_ = param_.getValue("missed_cleavages");
+  peptide_min_mass_ = param_.getValue("peptide_min_mass");
+  peptide_max_mass_ = param_.getValue("peptide_max_mass");
+  peptide_min_length_ = param_.getValue("peptide_min_length");
+  peptide_max_length_ = param_.getValue("peptide_max_length");
+  fragment_min_mz_ = param_.getValue("fragment_min_mz");
+  fragment_max_mz_ = param_.getValue("fragment_max_mz");
+  precursor_mz_tolerance_ = param_.getValue("precursor_mz_tolerance");
+  fragment_mz_tolerance_ = param_.getValue("fragment_mz_tolerance");
+  precursor_mz_tolerance_unit_ = param_.getValue("precursor_mz_tolerance_unit").toString();
+  fragment_mz_tolerance_unit_ = param_.getValue("fragment_mz_tolerance_unit").toString();
+  modifications_fixed_ = ListUtils::toStringList<std::string>(param_.getValue("modifications_fixed"));
+  modifications_variable_ = ListUtils::toStringList<std::string>(param_.getValue("modifications_variable"));
+  max_variable_mods_per_peptide_ = param_.getValue("max_variable_mods_per_peptide");
+}
+
+void SearchDatabase::search(MSSpectrum& spectrum, std::vector<SearchDatabase::Candidate>& candidates) const
+{ 
+  candidates.clear();
+  if (spectrum.size() == 0) return;
+  
+  unordered_set<size_t> index_hash; // saving every candidate only once
+
+  std::vector<Precursor> precursor = spectrum.getPrecursors();
+
+  if (precursor.size() != 1) return;
+
+  double prec_mz = precursor[0].getUnchargedMass();
+
+  spectrum.sortByIntensity(true);
+  size_t count_found = 0;
+  size_t count_rounds = 0;
+
+  auto prec_unit_da = precursor_mz_tolerance_unit_ == UNIT_DA;
+  auto frag_unit_da = fragment_mz_tolerance_unit_ == UNIT_DA;
+
+  for (const auto& peak : spectrum)
+  {
+    if (count_found == 0 && count_rounds == 5) return; // if peak with the 5 highest intensity don't match, skip whole spectrum
+
+    double new_frag_mz_tolerance = frag_unit_da ? fragment_mz_tolerance_ : ppmToMassAbs(fragment_mz_tolerance_, peak.getMZ());
+
+    auto itr_lower = lower_bound(bucket_frags_mz_.begin(), bucket_frags_mz_.end(), peak.getMZ() -  new_frag_mz_tolerance);
+    
+    size_t index_lower = distance(bucket_frags_mz_.begin(), itr_lower);
+
+    if (index_lower != 0) index_lower--; // lower bound returns one too high (because searching the minimum of each bucket)
+
+    while (bucket_frags_mz_[index_lower] <= (peak.getMZ() + new_frag_mz_tolerance) && index_lower < bucket_frags_mz_.size())
+    { 
+      auto bucket_start = all_fragments_.begin() + (index_lower * bucketsize_);
+      auto bucket_end = all_fragments_.begin() + (index_lower + 1) * bucketsize_;
+
+      auto condition = distance(all_fragments_.begin(), bucket_end) >= int(all_fragments_.size());
```

Comment:
maybe cast ptrdiff_t to size_t with abs or so, instead? You never know how many fragments you will end up having.

Reply:
also can you use a better variable name? which condition?